### PR TITLE
feat(lambda-python): add support for custom build image

### LINKED ISSUE #10298: [aws-lambda-python] Allow the use of CodeArtifact
Allow the use of CodeArtifact, so that private modules can be loaded from a private repository

### Use Case

Many librarys are private, which get used to build lambdas

### Proposed Solution

pass parameters for the code-artifact repo to the container. 

This is a :rocket: Feature Request

### PR DESCRIPTION
**Pull request by setu4993**
Adding support for custom build images for packaging Python dependencies. This resolves #10298 since Code Artifact login step with AWS CLI or credentials can be passed in with Docker build args, handled within the custom Docker build image.

This also makes https://github.com/aws/aws-cdk/pull/15306 redundant, so I closed it.

Fixes #10298, https://github.com/aws/aws-cdk/issues/12949 and https://github.com/aws/aws-cdk/issues/16234.

----

*By submitting this pull request, I confirm that my contribution is made under the terms of the Apache-2.0 license*

### CODE REVIEW

**Comment by setu4993** [comment]
@nija-at : This PR is ready for review.

**Comment by setu4993** [comment]
The CI test is failing on a test related to `appsync` for an API key error, which is weird since that's not something that this PR is touching. Not sure what's going on there...

**Comment by nija-at** [comment]
Thanks so much for submitting this pull request. I am marking this pull request as `p2`, which means that we are unable to work on it immediately, but it's definitely still on our radar.

We use +1s to help prioritize our work, and are happy to revaluate this pull request based on community feedback.

**Comment by setu4993** [comment]
Thanks, @nija-at. A few questions:
1. What do you think is the approximate timeframe for getting this in?
2. I posted on one of the issues a few weeks ago, but it has been mostly inactive. This is an important fix for us as we expand our use of aws-cdk, so if we can get this in sooner, that'd be very much appreciated.
3. This is already supported with the [`NodejsFunction`](https://docs.aws.amazon.com/cdk/api/latest/docs/@aws-cdk_aws-lambda-nodejs.BundlingOptions.html#dockerimage) and right now, the `PythonFunction` not supporting custom bundling keeps it at a lower par in comparison.

**Comment by setu4993** [comment]
@nija-at : Would appreciate any thoughts / comments about https://github.com/aws/aws-cdk/pull/15324#issuecomment-883059476 and this PR generally. Especially since without this PR, `PythonFunction` remains one step behind `NodeJsFunction` in supporting custom build images, which makes it less usable.

**Comment by SamStephens** [comment]
@setu4993 whilst I think this is great, I don't think it actually solves the Code Artifact issue https://github.com/aws/aws-cdk/issues/10298, because as discussed in https://github.com/aws/aws-cdk/issues/16234#issuecomment-929690728, Code Artifact credentials are emphemeral, they are only valid for 12 hours.

My understanding of Docker images is that they're relatively durable. I would expect that for a Function definition with `dockerBuildImageOptions` provided, a Docker image would be built based on the base SAM build image with the custom options applied, and then that Docker image would be used until either the base SAM build image changes, or the `dockerBuildImageOptions` change. Which doesn't work for something that is only valid for 12 hours.

**Comment by setu4993** [comment]
@SamStephens : Thanks for your feedback. The reason I think this solves https://github.com/aws/aws-cdk/issues/10298 is that Code Artifact credentials can be passed into the Docker image at the _build_ step. That allows installing private Python packages from Code Artifact, and then packaging the function to be executed. The credentials expiring later doesn't matter since the packages are already cached into the function once it has been built.

So, the idea would be that a custom `Dockerfile.build` contains steps / args required to either fetch Code Artifact credentials or parse them from build arguments (likely preferable).

We are already doing something similar for Fargate-based Docker images.

**Comment by setu4993** [comment]
@nija-at : Any additional feedback on timeline here? This also closes https://github.com/aws/aws-cdk/issues/16234, so there's atleast 2 issues open that could be resolved and use the changes in here.

File: packages/@aws-cdk/aws-lambda-python/README.md [hunk 1]
```diff
@@ -36,6 +36,32 @@ new PythonFunction(this, 'MyFunction', {
...
 ```
 
+Custom Docker images for bundling dependencies can be specified by specifying additional `buildImageOptions`. If using a custom Docker image, please ensure that the dependencies are stored at `/var/dependencies` within the Docker image for them to be bundled into the Lambda asset.
+
+A different bundling Docker image `Dockerfile.build` can be specified as:
+
+```ts
+new PythonFunction(this, 'MyFunction', {
+   ...
+   dockerBuildImageOptions: {
```
Comment (corymhall) [inline packages/@aws-cdk/aws-lambda-python/README.md:46]: @setu4993 I think that I would prefer to adopt the same approach that is used by both `NodejsFunction` and `GoFunction` and just allow the user to supply their own `DockerImage`. i.e.

```ts
new lambda.PythonFunction(this, 'my-handler', {
  bundling: {
    dockerImage: DockerImage.fromBuild('/path/to/Dockerfile'),
  },
});
```

**Review by corymhall — CHANGES_REQUESTED** [review]

File: packages/@aws-cdk/aws-lambda-python/README.md (hunk 1 above)
Comment (setu4993) [reply packages/@aws-cdk/aws-lambda-python/README.md:46]: Thanks for the feedback, @corymhall! I should have some space to work on this and update it over the weekend.

Comment (setu4993) [reply packages/@aws-cdk/aws-lambda-python/README.md:46]: @corymhall : Thinking through this a bit more, I'm open to switching to that structure, but I think it'd require quite a bit of refactoring of the current dependency installation system. I'm open to doing that (possibly in a separate PR first) and then inheriting those changes to support bundling here. Thoughts?

Comment (corymhall) [reply packages/@aws-cdk/aws-lambda-python/README.md:46]: @setu4993 If you are willing to work on that refactoring then I think it is a good idea! Before making this library stable we would want to bring it in line with the other two anyway.

Comment (setu4993) [reply packages/@aws-cdk/aws-lambda-python/README.md:46]: @corymhall : What do you think about https://github.com/aws/aws-cdk/pull/18082? The allure of getting `lambda-Python` to stable (finally!) was enough motivation :).

**Comment by setu4993** [comment]
Closing in favor of https://github.com/aws/aws-cdk/pull/18082.

**Pull request closed without merging**